### ISSUE TITLE
[Bug]: Deepseek V4 Flash 0731 BF16 weights converted by msmodelslim lose accuracy when `quant_model_description.json` is present

### ISSUE THREAD

**Issue by liyu119** [vllm-project/vllm-ascend#17476]
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Your output of above commands here
```

</details>

### 🐛 Describe the bug

**Branches checked**

- `vllm`: `releases/v0.26.0` @ `568afb3a13`
- `vllm-ascend`: `releases/v0.26.0rc` @ `16b0f0612`

The same bug is still present on these branches. Line numbers below refer to `vllm-ascend` `releases/v0.26.0rc`.

## Summary

BF16 checkpoints exported by msmodelslim ship a `quant_model_description.json` whose weight entries are all `"FLOAT"`. vLLM-Ascend treats the mere existence of that file as an Ascend quantization checkpoint and builds `AscendModelSlimConfig`. The GEMM / grouped-matmul kernels for layers that successfully skip quantization match the plain HF BF16 path. Accuracy still diverges on hash MoE layers whose `scoring_func` is `sqrtsoftplus`, because the FLOAT skip path drops `tid2eid`. Expert ids are then chosen from router logits only, so `npu_grouped_matmul` runs on a different expert set. Softmax / sigmoid layers do not read the table, which is why the mismatch shows up only on some layers and some requests.

Changing the dtype stored in the safetensors files does not leave this path. The loader never inspects weight dtypes. The shard filename (`quant_model_weights-*.safetensors` vs `model-*-of-*.safetensors`) is also not the switch.

## What is compared

| | Script export (matches reference) | msmodelslim convert (accuracy drift) |
| --- | --- | --- |
| Weights | `model-*-of-*.safetensors` | `quant_model_weights-*.safetensors` |
| Index | `model.safetensors.index.json` | `quant_model_weights.safetensors.index.json` |
| Descriptor | absent | `quant_model_description.json`, weight values all `"FLOAT"` |

### 1. The descriptor file forces the Ascend quant config

`detect_quantization_method` (`vllm_ascend/quantization/utils.py`, lines 114–117) returns `"ascend"` as soon as `quant_model_description.json` exists. It does not parse the JSON and does not open any safetensors file.

`maybe_auto_detect_quantization` (same file, lines 198–214), called from `NPUPlatform.check_and_update_config`, then sets `model_config.quantization = "ascend"` and rebuilds `vllm_config.quant_config` as `AscendModelSlimConfig`. With no descriptor, `quant_config` stays `None` and the log says the model is loaded as float.

The default weight loader globs `*.safetensors` and only consults `model.safetensors.index.json`. `quant_model_weights-*.safetensors` are still loaded. The filename difference is packaging, not the code-path switch.

### 2. `"FLOAT"` is a per-layer skip inside the quant config, not the HF path

`AscendModelSlimConfig.get_quant_method` (`vllm_ascend/quantization/modelslim_config.py`, lines 668–741) calls `is_layer_skipped_ascend` (lines 752–784). A layer is skipped only when the descriptor value is exactly the string `"FLOAT"` and the module prefix matches.

- Fused modules (`experts`, `qkv_proj`, `gate_up_proj`, …) require every shard key, e.g. `experts.0.gate_proj.weight`, `experts.0.up_proj.weight`, `experts.0.down_proj.weight`, to be `"FLOAT"`. A missing key is a `KeyError`, except the `k_eq_v` case that now skips a missing `v_proj` (`_is_missing_v_shard`, lines 765–766). Mixed FLOAT and a real quant type raises `ValueError`.
- Other linears skip when any descriptor key starts with the module prefix, ends with `.weight`, and equals `"FLOAT"`. If nothing matches, the layer is not skipped. `create_scheme_for_layer` then runs. There is no registered scheme for `"FLOAT"`, so that layer raises `NotImplementedError`, or it builds a real quant scheme if the value is something like `"W8A8_DYNAMIC"`.

When the skip hits, the kernels match the script path:

| Layer | FLOAT skip (prefix matches) | Not skipped |
| --- | --- | --- |
| Linear | `torch.ops.vllm.unquantized_gemm` → `F.linear` | `npu_dynamic_quant` + `npu_quant_matmul` (activation int8) |
| MoE | `npu_grouped_matmul` + `npu_swiglu` (`QuantType.NONE`) | `quant_apply_mlp`: `npu_dynamic_quant` + `grouped_matmul_swiglu_quant` |

On the script path, Ascend linear classes already use `AscendUnquantizedLinearMethod` when `quant_config is None`, and `AscendMoERunner` installs `AscendUnquantizedFusedMoEMethod` in that case (`vllm_ascend/ops/fused_moe/fused_moe.py`, lines 380–381). So a fully matched all-FLOAT descriptor does not by itself change the matmul operator. NZ conversion (`maybe_trans_nz`) is also shared by both unquantized methods and only runs for BF16/FP16 when `weight_nz_mode == 2`.

### 3. The accuracy bug: the FLOAT MoE skip drops `tid2eid`

`tid2eid` is an `int32` table of shape `[vocab_size, top_k]` mapping token id → expert id. DeepSeek V4 allocates it only for hash layers (`layer_idx < num_hash_layers` and not a draft layer) in `vllm_ascend/models/deepseek_v4.py` (lines 421–435) and passes it into `FusedMoE(..., tid2eid=self.gate.tid2eid)` (line 463). Other layers set it to `None`. Default `scoring_func` is `config.scoring_func` or `"softmax"` (line 451).

`AscendMoERunner` keeps the table on itself, then rebuilds `quant_method`:

```python
# fused_moe.py, lines 380–385
if routed_experts.quant_config is None:
    routed_experts.quant_method = AscendUnquantizedFusedMoEMethod(
        self.moe_config, tid2eid=self.tid2eid
    )
else:
    routed_experts.quant_method = routed_experts.quant_config.get_quant_method(
        routed_experts, self.layer_name, tid2eid=self.tid2eid
    )
```

The script path takes the first branch, so the method stores the table. The msmodelslim path takes the second. `get_quant_method` receives `tid2eid`, but the FLOAT skip does not forward it:

```python
# modelslim_config.py, lines 732–741
if self.is_layer_skipped_ascend(prefix, self.packed_modules_mapping):
    return AscendUnquantizedFusedMoEMethod(layer.moe_config)  # tid2eid dropped
return AscendFusedMoEMethod(scheme, layer.moe_config, tid2eid)  # real quant keeps it
```

The grouped matmul is driven by `quant_method.apply` (`AscendUnquantizedFusedMoEMethod.apply`, `fused_moe.py` lines 200–214), which calls `select_experts(..., tid2eid=self.tid2eid)`. After the FLOAT skip that argument is `None`. The runner's own `self.tid2eid` is still set, but it is not the table used for the expert ids that enter GMM.

`_select_experts_with_fusion_ops` (`vllm_ascend/ops/fused_moe/experts_selector.py`, lines 247–286) consumes the table only when `scoring_func == "sqrtsoftplus"`:

- `tid2eid is not None`: `torch.ops._C_ascend.moe_gating_top_k_hash` is called with `input_ids` and the table.
- `tid2eid is None`: the same op is called with `input_ids=None` and `tid2eid=None`, so experts come from logits only.
- `softmax` / `sigmoid` use `moe_gating_top_k` and ignore the table.
- If the fused gating op is unavailable, `select_experts` forces `tid2eid=None` (line 108) on both paths.

That is the intermittent behavior:

| Layer | Script (`quant_config is None`) | msmodelslim, all `"FLOAT"`, keys match | Result |
| --- | --- | --- | --- |
| Non-hash, or softmax / sigmoid | Table is unused | Table is `None` and still unused | Same |
| Hash and `sqrtsoftplus` | `moe_gating_top_k_hash` looks up `tid2eid` | Same op, empty table and empty `input_ids` | Different `topk_ids`; same `npu_grouped_matmul` runs on different experts |

A layer that is not skipped still receives `tid2eid` via `AscendFusedMoEMethod`. The drop is specific to the FLOAT skip.

### 4. Other descriptor-driven splits that can look like the same bug

These do not replace finding (3), but the same file can move individual layers onto different operators:

- Unmatched prefixes do not skip. Those linears use `npu_dynamic_quant` + `npu_quant_matmul`. Those MoE layers use `quant_apply_mlp` in `vllm_ascend/ops/fused_moe/moe_mlp.py`, which quantizes activations to int8. An all-FLOAT value with no registered scheme fails load instead.
- Non-weight descriptor fields still change the graph when weight entries are FLOAT. `norm.bias` adds a bias in `AscendRMSNorm`. `fa_quant_type` or `kv_cache_type == "C8"` changes the KV cache dtype and the attention kernel (`attention/attention_v1.py`, `modelslim_config.py` `_add_kvcache_quant_metadata`).
- MTP. `spec_decode/step3p5.py` attaches the verifier quant config to the draft. `quant_prefix_mapper` strips `.mtp_block.` for `step3p5_mtp` and renames `transformer.shared_head.*`. The script path does not. msmodelslim descriptors often omit MTP layers; a missing fused shard is a `KeyError`, and a present FLOAT shard falls into the `tid2eid` drop above.
- On this branch, `spec_decode/llm_base_proposer.py` `_load_quarot_rotation` (lines 410–440) opens `quant_model_description.json` whenever it exists. If `optional.quarot` is absent it looks for `optional/quarot.safetensors`, logs a warning, and returns `None` when that file is missing. An all-FLOAT descriptor without a rotation file does not rotate weights.

## Suggested fix

In `AscendModelSlimConfig.get_quant_method`, construct the skipped MoE method the same way `AscendMoERunner` does when `quant_config is None`:

```python
return AscendUnquantizedFusedMoEMethod(layer.moe_config, tid2eid=tid2eid)
```

Leave `AscendFusedMoEMethod(scheme, layer.moe_config, tid2eid)` as it is.

To match the script path without a code change, remove `quant_model_description.json`. Rewriting every descriptor value to `"FLOAT"`, or rewriting safetensors dtypes, still selects `quantization="ascend"`.

## Workaround

Delete `quant_model_description.json` from the msmodelslim directory (and do not pass `--quantization ascend`) so `detect_quantization_method` returns `None`.
